### ISSUE TITLE
Make the `$container` type more precise in case of guard conditions

### ISSUE THREAD

**Issue by JohannesMeierSE** [langium/langium#1300]
The following simple and artificial example uses [guard conditions](https://langium.org/docs/grammar-language/#guard-conditions) in order to switch between different cases (excerpt):

```langium
Teacher:
    'teacher' name=ID values=Reused<true>;

Student:
    'student' name=ID values=Reused<false>;

Reused<condition>:
    <!condition> OneValue | <condition> TwoValues;

OneValue:
    'values' value=ID;
TwoValues:
    'values' valueOne=ID valueTwo=ID;
```

([The whole grammar with an instance can be found in the playground.](https://langium.org/playground?grammar=OYJwhgthYgBA4gVxgEwMIHsB2AXMBLLAUxAFEAPSABwBsiAoeo3EAT1gFkMUiaAuerCGwAFDiJgAxgAsSAagC8AFQkySsAD6wAzjkQ9c2xQGU9BnAEoAVAG5GKqbJADhsAOTjHJN7CyQiCgCSACKwAG5gNIhE2goASkSI2kQoADw4INEAfHb0pvrMOC7CbroFuD5%2BEAEh4ZHRsQlJKakAZpHJOYxNyWmS2Cj4OPjYWcVCqQCE-ViDw6OwAPLEAGr1RJqwqTNzI1hZsEoA7hhrUTG5y0Rn0ePuEefaPg-RQcF2x6fr2nduLzHPdZXN51c6fN65aT4FAGWDiEAQQiRWAAdWMfFgAHoADpGTF2eGIvw0WAhDGYgDaAH0wABaABeAEFaQAtAC6FOxRypbKs%2BMYUJhzDhJCJyI4ABkqWhFhwOKQAHJKcnYnFWTnabHGXkAfmxVlV-MFsMJSJJxilMrliuVWNVqopAD1sVhsSBefygA&content=C4UwhgxgFiBOAEAVc07wG5gDYFcQGcNs8B5AOxCNxEQHcB7AKH2BwBMQzh4BlVjrlTyF6ZLAE9yIAGrEQjIA))

For this grammar, Langium generates the following types in the `ast.ts` (excerpt):

```ts
export type Reused = OneValue | TwoValues;

export interface OneValue extends AstNode {
    readonly $container: Student | Teacher;
    readonly $type: 'OneValue';
    value: string
}

export interface TwoValues extends AstNode {
    readonly $container: Student | Teacher;
    readonly $type: 'TwoValues';
    valueOne: string
    valueTwo: string
}
```

From my point of view, the types of the `$container` properties could be more precise, i.e. only `Student` for `OneValue` and only `Teacher` for `TwoValues`.
The current `$container` types are not wrong, but as a user of Langium, I would expect the following types, since the grammar clearly shows, that these types are enough:

```ts
export type Reused = OneValue | TwoValues;

export interface OneValue extends AstNode {
    readonly $container: Student;
    readonly $type: 'OneValue';
    value: string
}

export interface TwoValues extends AstNode {
    readonly $container: Teacher;
    readonly $type: 'TwoValues';
    valueOne: string
    valueTwo: string
}
```

**Comment by JohannesMeierSE** [comment]
If you think this issue is worth an improvement, I am willing to work on this issue.

**Comment by spoenemann** [comment]
Yes for such clear cases it would make sense to restrict the container type.
